### DISCUSSION TITLE
attributes for comparision

### DISCUSSION THREAD

**Discussion by jaigsingla** [pelicanmapping/osgearth#2254] [Support Q&A]
Dear all,

As we can get Text content  of shape attribute name and numerical expression in text-> priority. 

How to get attribute numeric values for comparison ?\

thanks,
Jay

**Comment by smarttigerfly** [comment]
I think you should use `script` like this:
```
const char* styles_css =
R"(
p{fill:#000080;fill-opacity: 0.5;stroke:#3f763f;stroke-width: 1px;altitude-clamping:terrain;altitude-technique:drape;}
p1:p{fill:#66FF66;}
p2:p{fill:#33CC33;}
p3:p{fill:#009900;}
)";

const char* script_source = R"(
function getStyleClass()
{
    var pop = parseFloat(feature.properties.pop);
    if ( pop<200000 ) {return "p1";}
    else if (pop < 500000) {return "p2";}
    else {return "p3";}
}
)";

int main(int argc, char** argv)
{
    osgEarth::initialize();
    osg::ArgumentParser arguments(&argc, argv);

    // create the map.
    osg::ref_ptr<Map> map = new Map();

    {
        GDALImageLayer* layer = new GDALImageLayer();
        layer->setURL("E:\\osgEarthData\\data\\world.tif");
        map->addLayer(layer);
    }


    // create a feature source to load the building footprint shapefile.
    OGRFeatureSource* data = new OGRFeatureSource();
    data->setName("buildings-data");
    data->setURL("E:\\osgEarthData\\css\\sld_cookbook_polygon.shp");
    Status s = data->open();
    if (s.isError())
    {
        OE_WARN << s.message() << std::endl;
        exit(0);
    }

    StyleSheet* stringSheet = new StyleSheet();
    stringSheet->addStylesFromCSS(styles_css);
    stringSheet->setScript(new StyleSheet::ScriptDef(script_source));
    stringSheet->addSelector(StyleSelector("default", StringExpression("getStyleClass()")));

    FeatureModelLayer* layer = new FeatureModelLayer();
    layer->setFeatureSource(data);
    layer->setStyleSheet(stringSheet);

    map->addLayer(layer);

    if (layer->getStatus().isError())
    {
        OE_WARN << layer->getStatus().message() << std::endl;
    }

    // initialize a viewer:
    osgViewer::Viewer viewer(arguments);

    EarthManipulator* manip = new EarthManipulator();
    viewer.setCameraManipulator(manip);

    osg::Group* root = new osg::Group();
    viewer.setSceneData(root);

    // make the map scene graph:
    MapNode* mapNode = new MapNode(map.get());
    root->addChild(mapNode);

    GLUtils::setGlobalDefaults(root->getOrCreateStateSet());

    viewer.realize();
    // zoom to a good startup position
    manip->setViewpoint(Viewpoint(
        "Home",
        0, 0, 0,   // longitude, latitude, altitude
        180, 0, 34500000.0), // heading, pitch, range
        5.0);                    // duration

    // This will mitigate near clip plane issues if you zoom in close to the ground:
    LogarithmicDepthBuffer buf;
    buf.install(viewer.getCamera());

    return viewer.run();
}
```

The shapefile `sld_cookbook_polygon.shp` has the `pop` attribute.

**Comment by jaigsingla** [comment]
int main(int argc, char** argv)
{
       osg::ArgumentParser arguments(&argc, argv);
       osg::DisplaySettings::instance()->setMinimumNumStencilBits(8);
      Map* map = new Map();
       // Add an imagery layer (blue marble from a TMS source)
       {
              GDALOptions gdal;
             gdal.url() = "G:/data/viz/world.tif";
              ImageLayer* baseLayer = new ImageLayer("world", gdal);
              map->addImageLayer(baseLayer);
       }
       osgViewer::Viewer viewer(arguments);
     
   //FeatureSource featureSrc;
       OGRFeatureOptions featureOptions;
       featureOptions.name() = "Shapefile";
       featureOptions.url() = "G:/viz/data/world.shp";

        osg::ref_ptr<FeatureSource> features = FeatureSourceFactory::create(featureOptions);
       osg::ref_ptr<FeatureCursor> cursor = features->createFeatureCursor();

       Style labelStyle;

       TextSymbol* text = labelStyle.getOrCreateSymbol<TextSymbol>();

       std::string g_priorityAttrib;

       g_priorityAttrib = "pop_cntry";

       text->content() = StringExpression("[cntry_name]");

       text->priority() = NumericExpression("[pop_cntry]");

       text->removeDuplicateLabels() = true;

        while (cursor->hasMore())
       {
              text->removeDuplicateLabels() = true;
              Feature* feature = cursor->nextFeature();
              float pop = feature->getDouble(g_priorityAttrib, 0.0);
                 if (pop > 10000000) {
                     text->fill()->color() = Color::White;
                     text->halo()->color() = Color::Blue;
                     text->size() = 18.0f;
              }
              else{
              text->size() = 16.0f;
              text->alignment() = TextSymbol::ALIGN_CENTER_CENTER;
              text->fill()->color() = Color::White;
              text->halo()->color() = Color::DarkGray;
              }
       }

       // and configure a model layer:

       FeatureGeomModelOptions geomOptions;

       geomOptions.featureOptions() = featureOptions;

       geomOptions.styles() = new StyleSheet();

       geomOptions.styles()->addStyle(labelStyle);

       ModelLayerOptions layerOptions("my features", geomOptions);

       MapNodeOptions mapNodeOptions;

       mapNodeOptions.enableLighting() = false;

       map->addModelLayer(new ModelLayer(layerOptions));

       MapNode* mapNode = new MapNode(map, mapNodeOptions);

      osgEarth::Util::EarthManipulator* manip = new EarthManipulator();

       root = new osg::Group();
       root->addChild(mapNode);
       viewer.setSceneData(root);
       viewer.setCameraManipulator(manip);
       //Viewer
   viewer.addEventHandler(new osgViewer::StatsHandler);
   viewer.addEventHandler(new osgViewer::WindowSizeHandler);
   viewer.addEventHandler(new osgViewer::ThreadingHandler());
       viewer.run();
       return 0;

thanks for the reply.. 

I was trying with above mentioned code. It is giving some access violation reading location error at cursor->hasMore

 Are you using osgearth 2.8 or later??

thanks 
Jay
